global_segment_map_node/src/controller.cpp

```diff
@@ -591,8 +706,92 @@ bool Controller::saveSegmentsAsMeshCallback(
   return overall_success;
 }
 
-bool Controller::extractInstancesCallback(std_srvs::Empty::Request& request,
-                                          std_srvs::Empty::Response& response) {
+bool Controller::getListSemanticInstancesCallback(
+    vpp_msgs::GetListSemanticInstances::Request& /* request */,
+    vpp_msgs::GetListSemanticInstances::Response& response) {
+  SemanticLabels semantic_labels;
+
+  {
+    std::lock_guard<std::mutex> label_tsdf_layers_lock(
+        label_tsdf_layers_mutex_);
+    // Get list of instances and their corresponding semantic category.
+    map_->getSemanticInstanceList(&response.instance_ids, &semantic_labels);
+  }
+
+  // Map class id to human-readable semantic category label.
+  for (const SemanticLabel semantic_label : semantic_labels) {
+    response.semantic_categories.push_back(classes[(unsigned)semantic_label]);
+  }
+  return true;
+}
+
+bool Controller::getAlignedInstanceBoundingBoxCallback(
+    vpp_msgs::GetAlignedInstanceBoundingBox::Request& request,
+    vpp_msgs::GetAlignedInstanceBoundingBox::Response& response) {
+  InstanceLabels all_instance_labels, instance_labels;
+  std::unordered_map<InstanceLabel, LabelTsdfMap::LayerPair>
+      instance_label_to_layers;
+  InstanceLabel instance_label = request.instance_id;
+
+  {
+    std::lock_guard<std::mutex> label_tsdf_layers_lock(
+        label_tsdf_layers_mutex_);
+    // Get list of all instances in the map.
+    all_instance_labels = map_->getInstanceList();
+  }
+  // Check if queried instance id is in the list of instance ids in the map.
+  auto instance_label_it = std::find(all_instance_labels.begin(),
+                                     all_instance_labels.end(), instance_label);
+
+  if (instance_label_it == all_instance_labels.end()) {
+    LOG(ERROR) << "The queried instance ID does not exist in the map.";
```

Comment:
It seems the ros service model doesn't have a canonical way of handling errors. However, it might still be nice to return something, right?